tensorflow/core/kernels/substr_op.cc

```diff
@@ -0,0 +1,145 @@
+        auto output = output_tensor->flat<string>();
+
+        // Set output to be substrings of input strings
+        for (size_t i = 0; i < input_tensor.NumElements(); ++i) {
+          OP_REQUIRES(context, pos >= 0 && pos < input(i).size(),
+                      errors::InvalidArgument("pos ", pos, 
+                                              " out of range for string b'", 
+                                              input(i), "' at index ", i));
+          output(i) = input(i).substr(pos, len);
+        }
+      } else if (input_shape == pos_shape) {
+        // Perform Op element-wise
+        auto input = input_tensor.flat<string>();
+        auto pos = pos_tensor.flat<T>();
+        auto len = len_tensor.flat<T>();
+
+        // Allocate output
+        Tensor* output_tensor = nullptr;
+        OP_REQUIRES_OK(context,
+                       context->allocate_output("output", input_tensor.shape(),
+                                                &output_tensor));
+        auto output = output_tensor->flat<string>();
+
+        // Set output to be substrings of input strings
+        for (size_t i = 0; i < input_tensor.NumElements(); ++i) {
+          OP_REQUIRES(context, pos(i) >= 0 && pos(i) < input(i).size(),
+                      errors::InvalidArgument("pos ", pos(i), 
+                                              " out of range for string b'", 
+                                              input(i), "' at index ", i));
+          output(i) = input(i).substr(pos(i), len(i));
+        }
+      } else {  
+        // Attempt broadcasting this Op
+        context->SetStatus(errors::Unimplemented(
+                 "Substr broadcast is not supported yet."));
+
+        // BCast bcast(BCast::FromShape(input_shape), BCast::FromShape(pos_shape));
+        // OP_REQUIRES(context, bcast.IsValid(), 
+        //             errors::InvalidArgument("Incompatible shapes: ", 
+        //                                     input_shape.DebugString(), " vs. ",
+        //                                     pos_shape.DebugString()));
+        // TensorShape output_shape = BCast::ToShape(bcast.result_shape());
+        // int ndims = output_shape.dims();
+        // Tensor* output_tensor = nullptr;
+        // OP_REQUIRES_OK(context,
+        //                context->allocate_output("output", output_shape,
+        //                                         &output_tensor));
+        // switch (ndims) {
+        //   case 2: {
+        //     auto output = output_tensor->shaped<string, 2>(bcast.x_reshape());
+        //     auto input_reshaped = input_tensor.shaped<string, 2>(bcast.x_reshape());
+        //     auto pos_reshaped = pos_tensor.shaped<T, 2>(bcast.y_reshape());
+        //     auto len_reshaped = len_tensor.shaped<T, 2>(bcast.y_reshape());
+              
+        //     typename TTypes<string, 2>::Tensor input = input_reshaped.broadcast(BCast::ToIndexArray<2>(bcast.x_bcast()));
```

Comment:
I have- my first attempts involved trying to use that code. The main problem I was encountering is that those end up using the Eigen::Matrix::binaryExpr function under-the-hood to do the actual computation. Since substring takes in three parameters (input string, start position of the substring, and length of substring), I was having trouble finding a clean way to pass in all of the necessary data using the BinaryOp pattern.